projects/gloo/pkg/servers/iosnapshot/history.go

```
@@ -162,40 +168,46 @@ func (h *historyImpl) getApiSnapshotSafe() *v1snap.ApiSnapshot {
 	return &clone
 }
 
-// appendKubeGatewayResources searches for routes and gateways
-// setting the returned list onto the current resource map at a given key or replacing the contents with an error
-func (h *historyImpl) appendKubeGatewayResources(ctx context.Context, resources map[string]interface{}) {
-	cli := h.getKubeGatewayClientSafe()
-
-	if cli == nil {
+// getKubeGatewayResources returns the list of resources specific to the Kubernetes Gateway integration.
+func (h *historyImpl) getKubeGatewayResources(ctx context.Context) ([]crdv1.Resource, error) {
+	kubeGatewayClient := h.getKubeGatewayClientSafe()
+	if kubeGatewayClient == nil {
 		// No client has been set, so the Kubernetes Gateway integration has not been enabled
-		return
+		return nil, nil
 	}
 
-	var routeList apiv1.HTTPRouteList
-	err := cli.List(ctx, &routeList)
-	httpRoutesKey := fmt.Sprintf("%s.%s", wellknown.GatewayGroup, wellknown.HTTPRouteKind)
-	if err != nil {
-		resources[httpRoutesKey] = err
-	} else {
-		resources[httpRoutesKey] = routeList
+	resources := []crdv1.Resource{}
+	gvks := []schema.GroupVersionKind{
+		wellknown.GatewayClassListGVK,
+		wellknown.GatewayListGVK,
+		wellknown.HTTPRouteListGVK,
+		wellknown.ReferenceGrantListGVK,
 	}
-
-	var gwList apiv1.GatewayList
-	err = cli.List(ctx, &gwList)
-	gwKey := fmt.Sprintf("%s.%s", wellknown.GatewayGroup, wellknown.GatewayKind)
-	if err != nil {
-		resources[gwKey] = err
-	} else {
-		resources[gwKey] = gwList
+	for _, gvk := range gvks {
+		list := &unstructured.UnstructuredList{}
+		list.SetGroupVersionKind(gvk)
+		err := kubeGatewayClient.List(ctx, list)
+		if err != nil {
+			return nil, err
+		}
+		for _, uns := range list.Items {
+			out := crdv1.Resource{}
+			err := runtime.DefaultUnstructuredConverter.FromUnstructured(uns.Object, &out)
```

What would cause to an error here? This implementation seems to change the behavior of the old API, so curious if it was intentional: in the past errors are surfaced to the user and the operation continues in a best-effort attempt. now errors will cause the operation to end early...is there a preferred approach either from a UX or product perspective?

---

looks like it might fail on unknown fields.

yeah i was thinking about what was the best way to return errors. since it's a flat list of resources now, i didn't want to just stick error strings in the list of results. also i'd almost rather fail loudly instead of quietly mixing errors in with the partial results. perhaps we can ask for field/product input if there's a preferred way to return the errors, and update it in a follow-up